### PR TITLE
No lock

### PR DESCRIPTION
**Pull request by jmarolf**
There are several situations where we lock inside of an async method when we do no need to.  Instead we can yield the thread until the lock in available

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs [hunk 1]
```diff
@@ -241,6 +245,18 @@ private static string GetDisplayName(ConfiguredProject configuredProject, Projec
                 projectData.DisplayName;
         }
 
+        private Task ExecuteWithinLockAsync(Action action) => _gate.ExecuteWithinLockAsync(JoinableCollection, JoinableFactory, action);
+
+        private int _isInitialized;
+
+        private async Task EnsureInitialized()
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs]: Would this be a good addition to `OnceInitializedOnceDisposedAsync`? Maybe push it down and make it `protected`, then remove the dupe from `UnconfiguredProjectContextProvider`

Also, should have `Async` suffix.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs [hunk 2]
```diff
@@ -241,6 +245,18 @@ private static string GetDisplayName(ConfiguredProject configuredProject, Projec
...
+        {
+            if (Interlocked.CompareExchange(ref _isInitialized, 1, 0) == 0)
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs]: Would `Interlocked.Exchange(ref _isInitialized, 1) == 0` be simpler?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs [hunk 3]
```diff
@@ -241,6 +245,18 @@ private static string GetDisplayName(ConfiguredProject configuredProject, Projec
...
+            {
+                await InitializeAsync();
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs]: Could just return the `Task` here, and `Task.CompletedTask` in the usual case (zero allocation).

**Comment by drewnoakes** [comment]
Potentially fixes #2714.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs (hunk 2 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs]: I ran a benchmark. `CompareExchange` is quite a bit faster.

| Method |        Mean |     Error |     StdDev | Median |
------------------------ |------------:|----------:|-----------:|-------:|
| Exchange | 158.0000 ns | 81.148 ns | 239.267 ns | 0.0 ns |
| CompareExchange_Match |  16.8293 ns | 16.653 ns |  44.162 ns | 0.0 ns |
| CompareExchange_NoMatch |   0.5814 ns |  1.393 ns |   3.790 ns | 0.0 ns |

```c#
[ClrJob]
public class ExchangeVsCompareExchangeBenchmarks
{
    private int _i;

    [IterationSetup]
    public void Setup() => _i = 0;

    [Benchmark]
    public void Exchange() => Interlocked.Exchange(ref _i, 1);

    [Benchmark]
    public void CompareExchange_Match() => Interlocked.CompareExchange(ref _i, 0, 1);

    [Benchmark]
    public void CompareExchange_NoMatch() => Interlocked.CompareExchange(ref _i, 1, 1);
}
```

```
BenchmarkDotNet=v0.11.1, OS=Windows 10.0.17763
Intel Xeon Silver 4110 CPU 2.10GHz, 2 CPU, 32 logical and 16 physical cores
  [Host] : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.7.3190.0
  Clr    : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.7.3190.0
```

Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs]: I am quite surprised by the results there actually. Let me dig a little more...

Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs]: Yeah those results above are garbage.

| Method |     Mean |     Error |    StdDev |   Median |
| ------------------------ |---------:|----------:|----------:|---------:|
| Exchange | 7.222 ns | 0.1440 ns | 0.4246 ns | 6.985 ns |
| CompareExchange_Match | 9.680 ns | 0.1886 ns | 0.2880 ns | 9.555 ns |
| CompareExchange_NoMatch | 7.586 ns | 0.0116 ns | 0.0109 ns | 7.587 ns |

```c#
[ClrJob]
public class ExchangeVsCompareExchangeBenchmarks
{
    private const int OperationCount = 100_000_000;

    private int _i;

    [Benchmark(OperationsPerInvoke = OperationCount * 2)]
    public void Exchange()
    {
        _i = 0;

        for (var i = 0; i < OperationCount; i++)
        {
            Interlocked.Exchange(ref _i, 0);
            Interlocked.Exchange(ref _i, 1);
        }
    }

    [Benchmark(OperationsPerInvoke = OperationCount * 2)]
    public void CompareExchange_Match()
    {
        _i = 0;

        for (var i = 0; i < OperationCount; i++)
        {
            Interlocked.CompareExchange(ref _i, 0, 1);
            Interlocked.CompareExchange(ref _i, 1, 0);
        }
    }

    [Benchmark(OperationsPerInvoke = OperationCount * 2)]
    public void CompareExchange_NoMatch()
    {
        _i = 0;

        for (var i = 0; i < OperationCount; i++)
        {
            Interlocked.CompareExchange(ref _i, 1, 1);
            Interlocked.CompareExchange(ref _i, 1, 1);
        }
    }
}
```

There's not much in it. I'd go for `Exchange` based on it being the simpler form, but it's not a biggie. My curiousity's been satisfied.

**Comment by drewnoakes** [comment]
Looking at the source of `OnceInitializedOnceDisposedAsync.InitializeAsync`, I'm thinking it already handles the 'ensure' logic. Maybe the extra layer is redundant. The docs suggest as much:

> Initializes this instance if it has not yet been.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Utilities/EnsureOnceInitializedOnceDisposedAsync.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.ProjectSystem
+{
+    internal abstract class EnsureOnceInitializedOnceDisposedAsync : OnceInitializedOnceDisposedAsync
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Utilities/EnsureOnceInitializedOnceDisposedAsync.cs]: This isn't needed, OnceInitializedOnceDisposedAsync already gives this guarantee.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs [hunk 5]
```diff
@@ -64,7 +67,7 @@ public void ReleaseProjectContext(AggregateCrossTargetProjectContext context)
         {
             Requires.NotNull(context, nameof(context));
 
-            lock (_gate)
+            using (_gate.DisposableWait())
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs]: Looks like you can use the async form here and flow the task backwards.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs [hunk 6]
```diff
@@ -146,15 +150,15 @@ private ProjectData GetProjectData()
...
             {
                 return _configuredProjectContextsMap.TryGetValue(configuredProject, out targetedProjectContext);
             }
         }
 
         private void AddConfiguredProjectState(ConfiguredProject configuredProject, ITargetedProjectContext projectContext)
         {
-            lock (_gate)
+            using (_gate.DisposableWait())
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs]: Ditto (async form).

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs [hunk 7]
```diff
@@ -146,15 +150,15 @@ private ProjectData GetProjectData()
 
         private bool TryGetConfiguredProjectState(ConfiguredProject configuredProject, out ITargetedProjectContext targetedProjectContext)
         {
-            lock (_gate)
+            using (_gate.DisposableWait())
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs]: Ditto (async) via `Task<(bool found, ITargetedProjectContext targetedProjectContext)>` or similar.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs [hunk 8]
```diff
@@ -284,7 +267,7 @@ protected override async Task DisposeCoreAsync(bool initialized)
 
         private void DisposeAndClearSubscriptions()
         {
-            lock (_linksLock)
+            using (_linksLock.DisposableWait())
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs]: Ditto (async).

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/UnconfiguredProjectContextProvider.cs [hunk 9]
```diff
@@ -5,10 +5,12 @@
...
 using System.ComponentModel.Composition;
 using System.IO;
+using System.Threading;
 using System.Threading.Tasks;
 
 using Microsoft.VisualStudio.LanguageServices.ProjectSystem;
 using Microsoft.VisualStudio.ProjectSystem.LanguageServices;
+using Microsoft.VisualStudio.Threading;
 
 namespace Microsoft.VisualStudio.ProjectSystem.VS.LanguageServices
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/UnconfiguredProjectContextProvider.cs]: Not sure it's worth changing stuff in the language service. @davkean may say otherwise.

**Comment by jmarolf** [comment]
@drewnoakes in a separate PR we can look at making all methods async and not blocking any of the calling threads.  If I were to do this it would change about ~1000 methods, some of which are part of the public API.  This PR focuses on the less controversial fix of not blocking a thread when a method is already async. 

**NOTE:** you cannot use `DisposableWaitAsync` safely.  You need to instead use `ExecuteWithinLockAsync` in order to prevent deadlocks.

https://github.com/dotnet/project-system/blob/b72d88016d51f59cebca83b84428d494c66aaa11/src/Microsoft.VisualStudio.ProjectSystem.Managed/Threading/SemaphoreSlimExtensions.cs#L41-L57

**Comment by drewnoakes** [comment]
No worries if you want to do it in a separate PR&mdash;this is already a big improvement in reduction of blocking. I'd be keen to see you merge this sooner anyway, as I have some overlapping changes inbound.

> it would change about ~1000 methods, some of which are part of the public API

In my quick look at this I felt the ones I highlighted would require asyncifying fewer than 10 (only one or two levels back), none of which are public.

And yes, the goal was to replace the blocking `DisposableWait` with `ExecuteWithinLockAsync`, even if the func is synchronous. No one ends up in a blocking wait.

**Comment by jmarolf** [comment]
@davkean 	any problems here I should know about?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs (hunk 5 above)
Comment (jmarolf) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/AggregateCrossTargetProjectContextProvider.cs]: I've added async versions for these. As you said, it didn't end up spreading the virus very far :)

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs [hunk 10]
```diff
@@ -216,9 +201,9 @@ private Task<AggregateCrossTargetProjectContext> UpdateProjectContextAsync()
             });
         }
 
-        private void DisposeAggregateProjectContext(AggregateCrossTargetProjectContext projectContext)
+        private async Task DisposeAggregateProjectContextAsync(AggregateCrossTargetProjectContext projectContext)
         {
-            _contextProvider.Value.ReleaseProjectContext(projectContext);
+            await _contextProvider.Value.ReleaseProjectContextAsync(projectContext);
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs]: Unneeded state machine, just return.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs [hunk 11]
```diff
@@ -227,9 +212,9 @@ private async Task AddSubscriptionsAsync(AggregateCrossTargetProjectContext newP
 
             await _commonServices.ThreadingService.SwitchToUIThread();
 
-            await _tasksService.LoadedProjectAsync(() =>
+            await _tasksService.LoadedProjectAsync(async () =>
             {
-                lock (_linksLock)
+                await _linksLock.ExecuteWithinLockAsync(JoinableCollection, JoinableFactory, () =>
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs]: Unneeded state machine, just return.

**Comment by davkean** [comment]
I'm unconvinced by some of these, they basically have zero/very little contentention, don't continue on another thread and we've introduced JoinableTask overhead. We've probably made some of these situations worse as some of them just update a field, or add something to a collection - it's probably going to take longer for the async plumbing to schedule a continuation than its going to unblock the thread. I unconvinced that thread pool usage won't go up running all the continuations.

Convince me why it a good thing to replace all locks with async machinery?

**Comment by jmarolf** [comment]
## TL;DR

Collected some traces for this and it looks like this change is not worth it.  Thanks @davkean for challenging me on this. Below is the record of what I did, hopefully this is helpful to someone in the future.

## Process
### Setup
1. Enlist in `dotnet/project-system`
2. Enlist in `dotnet/roslyn`
3. install latest lab/d16.0stg build of VS

### Collection
For each branch (master and no-lock)
1. checkout project-system branch
2. build release version (`build.cmd /release /rootSuffix Exp`)
3. sync roslyn repo to [this commit](https://github.com/dotnet/roslyn/commit/547fbc0e803540c3b494e927ae680894cd30f2ef)
4. run `git clean -xdf` in the roslyn repo
5. run `Restore.cmd` in the roslyn repo
6. Start perf trace 
`>PerfView64.exe collect C:\perfTrace\Trace.etl -CircularMB:4096 -BufferSizeMB:256 -Merge:true -Zip:true -Providers:641d7f6c-481c-42e8-ab7e-d18dc5e5cb9e,*Microsoft-VisualStudio-Common,.NETTasks:0:0 -ThreadTime -NoV2Rundown -NoNGenRundown`
7. start VS `devenv /rootSuffix Exp C:\source\roslyn\Roslyn.sln`
8. Wait for all the nodes in the project system tree to not have yellow triangles
9. Stop perf collection

### Analysis
1. Find the *start* timestamp by looking at the process start event time
2. Find the *stop* event time by finding the last tree update event that was fired

**NOTE:** Traces can be found here: \\mlangfs1\public\jmarolf\traces

## Results
|                            |    [master](https://github.com/dotnet/project-system/commit/b713612310d72c45476c40b9b0ff9b5603c47e55)     |  [jmarolf:no-lock](https://github.com/jmarolf/project-system/commit/812dd253c5a7d9f7b5a1a28ccb42ff4a7b467cb9) |  Delta   |
|----------------------------|:-------------:|:----------------:|:--------:|
| Start (ms)                 | 1,751.522     |  723.096         |          |
| Stop (ms)                  | 41,743.366    |  40,316.409      |          |
| Elapsed Time (ms)          | 39,991.844    |  39,593.313      | 1.00%    |
| CPU Stack Time Sum (ms)    | 188,536.20    |  222,953.30      | -18.25%  |
| Thread Time Stacks Sum (ms)| 188,536.20    |  222,953.30      | -57.96%  |
| # Threads                  | 118           |  117             | 0.85%    |
| BLOCKED_TIME Sum (ms)      | 2,823,503.50  |  4,514,176.50    | -59.88%  |
| BLOCKED_TIME Count         | 134,592       |  382,517         | -184.20% |
| CPU_TIME Sum (ms)          | 173,050.13    |  222,757.33      | -28.72%  |
| CPU_TIME Count             | 330,200       |  605,195         | -83.28%  |
| READIED_TIME Sum (ms)      | 4,194.45      |  5,129.56        | -22.29%  |
| READIED_TIME Count         | 2,493         |  3,255           | -30.57%  |
